controller/dataplane/watch.go

```diff
@@ -27,5 +36,58 @@ func DataPlaneWatchBuilder(mgr ctrl.Manager) *builder.Builder {
 		// Watch for changes in HPA created by the dataplane controller.
 		Owns(&autoscalingv2.HorizontalPodAutoscaler{}).
 		// Watch for changes in PodDisruptionBudgets created by the dataplane controller.
-		Owns(&policyv1.PodDisruptionBudget{})
+		Owns(&policyv1.PodDisruptionBudget{}).
+		// Watch for changes in ConfigMaps created by the dataplane controller.
+		Owns(&corev1.ConfigMap{}).
+		// Watch for changes in ConfigMaps that are mapped to KongPluginInstallation objects.
+		// They may trigger reconciliation of DataPlane resources.
+		WatchesRawSource(
+			source.Kind(
+				mgr.GetCache(),
+				&corev1.ConfigMap{},
+				handler.TypedEnqueueRequestsFromMapFunc(listDataPlanesReferencingKongPluginInstallation(mgr.GetClient())),
+			),
+		)
+}
+
+func listDataPlanesReferencingKongPluginInstallation(
+	c client.Client,
+) handler.TypedMapFunc[*corev1.ConfigMap, reconcile.Request] {
+	return func(
+		ctx context.Context, kpiCM *corev1.ConfigMap,
+	) []reconcile.Request {
+		logger := ctrllog.FromContext(ctx)
+
+		// Find all DataPlane resources referencing KongPluginInstallation
+		// that maps to the ConfigMap enqueued for reconciliation.
+		kpiToFind := kpiCM.Annotations[consts.AnnotationMappedToKongPluginInstallation]
+		if kpiToFind == "" {
+			return nil
+		}
+
+		var dataPlaneList operatorv1beta1.DataPlaneList
+		if err := c.List(ctx, &dataPlaneList); client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "Failed to list DataPlanes in watch", "KongPluginInstallation", kpiToFind)
```

Comment:
This could also use an index similarly as we're doing it in Konnect entities e.g. https://github.com/Kong/gateway-operator/blob/4731ea64685befe828bbdcdc648d6f5e11660271/controller/konnect/index_kongpluginbinding.go right? Then there's no need for filtering below and listing is faster.

WDYT?

Reply:
It might be indeed lower gain than for the `Consumer`s but I believe it's still worth to add this index.

Can be done in a separate PR since this one is pretty big already.

Reply:
Ok, I'll address it separately and check for others opportunities